### DISCUSSION TITLE
Wrong font rendering on custom border characters

### DISCUSSION THREAD

**Discussion by seblyng** [ghostty-org/ghostty#3432] [Feature Requests, Ideas]
Hi :)

# Problem

I have created a custom font of the rounded border characters to make them "filled". I really like this look, as it makes popups more "modern" in my eyes.

This works great in kitty, but doesn't work as expected in ghostty.

## Kitty:
<img width="219" alt="Image" src="https://github.com/user-attachments/assets/7ddcd308-f1ad-4427-a947-509edf6b2ab2" />

## Ghostty:
<img width="215" alt="Image" src="https://github.com/user-attachments/assets/51022aab-29ca-487a-8f79-b959923b2bd9" />

# General info

I am using macOS 15.1.1 (24B91) and this is the output of `ghostty --version`:
```
Ghostty 1.0.0

Version
  - version: 1.0.0
  - channel: stable
Build Config
  - Zig version: 0.13.0
  - build mode : builtin.OptimizeMode.ReleaseFast
  - app runtime: apprt.Runtime.none
  - font engine: font.main.Backend.coretext
  - renderer   : renderer.Metal
  - libxev     : main.Backend.kqueue
```

# Root cause

Debugging this, I think I found out why this happens, but I am not sure about how to potentially resolve this. Here: https://github.com/ghostty-org/ghostty/blob/a8e5eef11cc67f87f445626f9ca2993373774bf8/src/font/sprite/Box.zig#L251 there seems to be special handling on certain unicodes and how to draw them, and since I have created a custom font and just used some random unicode points (U+e4ce-U+e4d4) they are not handled specially, and therefore not rendered as expected...

I noticed this when trying to copy the rounded border characters to the unicode value I am using, and noticed that they are rendered differently. I also tried to copy my unicode and use some of the characters that are properly handled, and they seem to be rendered as expected then.

However, I don't want to override some very normal border unicode characters.

This is how my border characters look:
 <img width="349" alt="Image" src="https://github.com/user-attachments/assets/651e8c0c-9e8e-488f-8ea9-994070fd6f45" />

and you can find the font here: https://github.com/seblj/dotfiles/blob/master/fonts/corners.ttf

**Comment by qwerasd205** [comment]
https://ghostty.org/docs/config/reference#font-codepoint-map

**Comment by seblyng** [comment]
No, that is not the issue. The issue is that the border characters are not rendered specially like the other "normal" border characters.

The font is correctly used

**Comment by qwerasd205** [comment]
Is the issue that they're the wrong size? Perhaps you would want to adjust the cell size down (there's config options for this) so they fit exactly from your font?

**Comment by seblyng** [comment]
Yeah maybe. But the same issue is if you for example copy the border characters from U+256d-U2570 and put them in for example U+e400-U+e403, then they will get rendered incorrectly because the switch never happens

**Comment by qwerasd205** [comment]
Ghostty only does custom rendering for known codepoints which need to fill the cell in some way. In the future there may be a way to configure certain codepoints to be automatically stretched to the cell width/height, but that is not currently possible.

**Comment by mitchellh** [comment]
Moving to issue triage until we're sure if this is Q&A or an issue...

**Comment by seblyng** [comment]
But this means that there is no workaround for this as of now?

I tried to patch it with something like this just to make it work:
```diff
diff --git a/src/font/sprite/Box.zig b/src/font/sprite/Box.zig
index ba7caa26..22f2b72f 100644
--- a/src/font/sprite/Box.zig
+++ b/src/font/sprite/Box.zig
@@ -247,6 +247,22 @@ pub fn renderGlyph(
 fn draw(self: Box, alloc: Allocator, canvas: *font.sprite.Canvas, cp: u32) !void {
     _ = alloc;
     switch (cp) {
+        // '╭'
+        0xe4ce => try self.draw_arc(canvas, .br, .light),
+        // '╮'
+        0xe4cf => try self.draw_arc(canvas, .bl, .light),
+        // '╯'
+        0xe4d0 => try self.draw_arc(canvas, .tl, .light),
+        // '╰'
+        0xe4d1 => try self.draw_arc(canvas, .tr, .light),
+        // '─'
+        0xe4d2 => self.draw_lines(canvas, .{ .left = .light, .right = .light }),
+        0xe4d3 => self.draw_lines(canvas, .{ .left = .light, .right = .light }),
+
+        // '│'
+        0xe4d5 => self.draw_lines(canvas, .{ .up = .light, .down = .light }),
+        0xe4d4 => self.draw_lines(canvas, .{ .up = .light, .down = .light }),
+
         // '─'
         0x2500 => self.draw_lines(canvas, .{ .left = .light, .right = .light }),
         // '━'
```

but it seems like it never enters the function for those characters...

I know this is not the "correct" way to fix this, but I just wanted to see if this would work

**Reply by qwerasd205** [reply]
The workaround for now is to adjust your cell size down to fit your characters. And yes, that patch would not work because there are other places prior to that function that filter down the codepoints that get sent to it.

**Reply by seblyng** [reply]
Doesn't seem like https://ghostty.org/docs/config/reference#adjust-cell-width and height helps if it were those two options you were thinking about with adjusting the cell size

**Comment by seblyng** [comment]
I am trying to modify my font to make it work, but I am unable to make it work as I want because I cannot seem to combine two unicode characters and not have any gap between them:

This is how it looks in kitty:
<img width="384" alt="image" src="https://github.com/user-attachments/assets/a09f0b38-01ba-4aa9-8ed6-51836a93b2f3" />


And this is ghostty:
<img width="396" alt="image" src="https://github.com/user-attachments/assets/e73617f6-41da-4aee-9b0f-bfdfc19e4577" />

There is a small gap between the characters. I am not any expert with fonts and unicodes, but I have used fontforge to make this icon and set the width to 779 with left and right bearings of -50 to try and have it overlap.

I have tried to experiment with `adjust-cell-width` to make it work, but can't seem to make it work.

**Comment by askorupskyy** [comment]
In my case this happened as well. Idk what the problem was but apparently iTerm2 and Ghostty rendered the same character differently. I checked the following conditions:

- Same font size
- Same font
- Same scaling
- Etc

The issue turned out to be with the font itself: it didn't support that symbol, so I just replaced it with the supported one (it was falling back onto the system default). Still don't know why it renders the characters differently, but it resolved my problem. 

It is still rendering the icons differently, but at least looks fine. (left iterm2, right ghostty)

<img width="552" alt="Screenshot 2024-12-28 at 12 10 06 AM" src="https://github.com/user-attachments/assets/10fd0b09-6992-4c47-bfde-20037f6442f9" />
